File: app/src/main/java/com/hover/stax/onboarding/OnBoardingActivity.java
package com.hover.stax.onboarding;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.hover.stax.R;
import com.hover.stax.databinding.OnboardingLayoutBinding;
import com.hover.stax.home.MainActivity;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.utils.Utils;


public class OnBoardingActivity extends AppCompatActivity implements ViewPager.OnPageChangeListener, View.OnClickListener {
    private static final String TAG = "OnboardingActivity";

    private static final int FIRST_SCROLL_DELAY = 4000;
    private static final int OTHER_SCROLL_DELAY = 5000;
    private static final double SWIPE_DURATION_FACTOR = 2.0;
    private static final double AUTOSCROLL_EASE_DURACTION_FACTOR = 5.0;

    private StaxAutoScrollViewPager viewPager;

    private OnboardingLayoutBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        UIHelper.setFullscreenView(this);
        super.onCreate(savedInstanceState);
        Utils.logAnalyticsEvent(getString(R.string.visit_screen, getString(R.string.visit_onboarding)), this);
        binding = OnboardingLayoutBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setupOnboardingSlides();
        initContinueButton();
    }

    private void setFullscreenView() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            getWindow().setDecorFitsSystemWindows(false);
        } else {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_FULLSCREEN);
        }
    }
    private void initContinueButton() {
        binding.onboardingContinueBtn.setOnClickListener(this);
    }


    private void setupOnboardingSlides() {
        viewPager = binding.vpPager;
        FragmentPagerAdapter adapterViewPager = new SlidesPagerAdapter(getSupportFragmentManager(), FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        viewPager.startAutoScroll(FIRST_SCROLL_DELAY);
        viewPager.setInterval(OTHER_SCROLL_DELAY);
        viewPager.setCycle(true);
        viewPager.setAutoScrollDurationFactor(AUTOSCROLL_EASE_DURACTION_FACTOR);
        viewPager.setSwipeScrollDurationFactor(SWIPE_DURATION_FACTOR);
        viewPager.setStopScrollWhenTouch(true);
        viewPager.addOnPageChangeListener(this);
        viewPager.setAdapter(adapterViewPager);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
        Utils.logAnalyticsEvent(getString(R.string.viewing_onboarding_slide, String.valueOf(position)), this);
    }

    @Override
    public void onPageScrollStateChanged(int state) {
    }

    @Override
    public void onClick(View v) {
        viewPager.stopAutoScroll();
        Utils.logAnalyticsEvent(getString(R.string.clicked_getstarted), this);
        setPassedThrough();
        goToMainActivity();
    }

    private void goToMainActivity() {
        startActivity(new Intent(this, MainActivity.class).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP));
        finish();
    }

    public static boolean hasPassedThrough(Context context) {
        return Utils.getBoolean(TAG, context);
    }

    private void setPassedThrough() {
        Utils.saveBoolean(TAG, true, this);
    }
}

File: app/src/main/java/com/hover/stax/onboarding/OnBoardingActivity.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.onboarding

import android.content.Intent
import android.os.Bundle
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import com.hover.sdk.permissions.PermissionHelper
import com.hover.stax.FRAGMENT_DIRECT
import com.hover.stax.OnboardingNavigationDirections
import com.hover.stax.R
import com.hover.stax.databinding.OnboardingLayoutBinding
import com.hover.stax.home.MainActivity
import com.hover.stax.home.NAV_HOME
import com.hover.stax.home.NAV_LINK_ACCOUNT
import com.hover.stax.login.AbstractGoogleAuthActivity
import com.hover.stax.permissions.PermissionUtils
import com.hover.stax.utils.AnalyticsUtil
import com.hover.stax.utils.NavUtil
import com.hover.stax.utils.UIHelper
import com.hover.stax.utils.Utils

class OnBoardingActivity : AbstractGoogleAuthActivity() {

    private lateinit var binding: OnboardingLayoutBinding
    private lateinit var navController: NavController

    override fun onCreate(savedInstanceState: Bundle?) {
        UIHelper.setFullscreenView(this)
        super.onCreate(savedInstanceState)

        logEvents()
        binding = OnboardingLayoutBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupNavigation()
        navigateNextScreen()
    }

    private fun setupNavigation() {
        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_fragment_onboarding) as NavHostFragment
        navController = navHostFragment.navController
    }

    private fun logEvents() = AnalyticsUtil.logAnalyticsEvent(getString(R.string.visit_screen, getString(R.string.visit_onboarding)), this)

    private fun navigateNextScreen() {
        if (hasPassedOnboarding()) checkPermissionsAndNavigate()
        else NavUtil.navigate(navController, OnboardingNavigationDirections.toWelcomeFragment())
    }

    fun checkPermissionsAndNavigate() {
        val permissionHelper = PermissionHelper(this)
        if (permissionHelper.hasBasicPerms()) navigateToMainActivity()
        else showBasicPermission()
    }

    private fun showBasicPermission() {
        PermissionUtils.showInformativeBasicPermissionDialog(
            0,
            { PermissionUtils.requestPerms(NAV_HOME, this) }, {
            AnalyticsUtil.logAnalyticsEvent(getString(R.string.perms_basic_cancelled), this)
        }, this
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        PermissionUtils.logPermissionsGranted(grantResults, this)
        checkPermissionsAndNavigate()
    }

    private fun navigateToMainActivity() {
        val intent = Intent(this, MainActivity::class.java).apply {
            putExtra(FRAGMENT_DIRECT, NAV_LINK_ACCOUNT)
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        }
        startActivity(intent)

        setPassedOnboarding()
        finish()
    }

    private fun setPassedOnboarding() = Utils.saveBoolean(OnBoardingActivity::class.java.simpleName, true, this)

    private fun hasPassedOnboarding(): Boolean = Utils.getBoolean(OnBoardingActivity::class.java.simpleName, this)
}